### DISCUSSION TITLE
Wallet Connect Unreal Engine 5

### DISCUSSION THREAD

**Discussion by shkvik** [WalletConnect/walletconnect-monorepo#1668] [Developer Support] [answered]
Wallet Connect Unreal Engine 5
===========

I need help with developing a WalletConnect client for unreal engine 5. I disassembled the protocol version 1, tested it through a ready- made [WalletConnectSharp](https://github.com/WalletConnect/WalletConnectSharp) client and ran into a socket response problem. I was checking the HMAC-256 authentication method defined in the [WalletConnectSharp[AESCipher.cs]](https://github.com/WalletConnect/WalletConnectSharp/blob/main/WalletConnectSharp.Core/Network/AESCipher.cs) repository. Hello @ecp4224.

I also successfully tested the decryption using the AES-256-CBC encryption algorithm with PKCS7 padding. I had to write from scratch the entire implementation of packaging encrypted values in a json rpc request, because at the moment there is no implementation of WalletConnect in c++. 

Encryption and decryption testing was successful, but I still do not receive a response from the WalletConnect socket bridges, this time I want to ask the developers to help me. Maybe I'm missing some important detail. I have an assumption that perhaps I am not getting a response because of the special format of the WebSocket headers or because I do not fall into the connection time interval. But I tried sending the generated data to WalletConnectSharp from other sockets, and it worked. This baffles me.

I will write the data that my code generates and the WalletConnectSharp code for comparison. I really don't understand why I can't get a response from your server. And yes, I know that version 1 will not be supported soon, but I do not plan to stay on the first version and it is important for me to understand exactly what format your system needs in order to successfully accept and respond to requests.

**My sequence for connection:**
- [X] Open connection with wss://(random).bridge.walletconnect.org
- [X] Subscribe to JSON RPC Response
- [X] Publish to JSON RPC Response
- [ ] Received request

**Сontent**
1. [QR Code URI](#QRCodeURI)  
2. [Subscribe](#Subscribe)
3. [Wallet Connect Session Request](#WalletConnectSessionRequest)
4. [AES-256-CBC PKCS7](#AES256CBC)
5. [Publish](#Publish)

<a name="QRCodeURI"/>

## QR Code URI 

 🟢  Generated data, using code written by me.
```c++
QR Code URI = "wc:f658bb91-4d4b-1526-0f20-578c4bda0f53@1?bridge=https%3A%2F%2Fu.bridge.walletconnect.org&key=edb61c40d8bb1afbde46ea8d2aa2c86eec2ec6cf494c8e5bf9bfca9693e25d85";
WSS URL = "wss://u.bridge.walletconnect.org";
Client id = "1fa6d81a-4888-966f-d0f9-29a3f45971ce";
Handshake Topic = "f658bb91-4d4b-1526-0f20-578c4bda0f53";

Bridge = "u";
```
🔵   Generated data, using code written [WalletConnectSharp](https://github.com/WalletConnect/WalletConnectSharp)
```c++
QR Code URI = "wc:caf40d9a-387d-4bff-8e31-0bbedefae326@1?bridge=https%3A%2F%2Fc.bridge.walletconnect.org&key=8185508e7f7e4d9b2df5778a0b245a46a5f65c9b8b7184f0c7ea6de382dc0b6f";
WSS URL = "wss://c.bridge.walletconnect.org";
Client id = "9444e269-007b-4f62-a87e-5d7c9a8f9cb3";
Handshake Topic = "caf40d9a-387d-4bff-8e31-0bbedefae326";

Bridge = "c";
```

<a name="Subscribe"/>

## Subscribe
 🟢  Generated data, using code written by me.
```json
{
  "topic": "f658bb91-4d4b-1526-0f20-578c4bda0f53",
  "type": "sub",
  "payload": "",
  "silent": true
}
```
🔵   Generated data, using code written [WalletConnectSharp](https://github.com/WalletConnect/WalletConnectSharp)
```json
{
  "topic": "9444e269-007b-4f62-a87e-5d7c9a8f9cb3",
  "type": "sub",
  "payload": "",
  "silent": true
}
```

<a name="WalletConnectSessionRequest"/>

## Wallet Connect Session Request
 🟢  Generated data, using code written by me.
```json
{
  "params": [
    {
      "peerId": "1fa6d81a-4888-966f-d0f9-29a3f45971ce",
      "peerMeta": {
        "description": "An example that showcases how to use the WalletConnectSharp library",
        "url": "https://app.warriders.com/",
        "icons": [
          "https://app.warriders.com/favicon.ico"
        ],
        "name": "WalletConnectUnreal"
      },
      "chainId": 1
    }
  ],
  "id": 1669068397634322,
  "jsonrpc": "2.0",
  "method": "wc_sessionRequest"
}
```
🔵   Generated data, using code written [WalletConnectSharp](https://github.com/WalletConnect/WalletConnectSharp) in [WcSessionRequest](https://github.com/WalletConnect/WalletConnectSharp/blob/main/WalletConnectSharp.Core/Models/WCSessionRequestRequest.cs)

```json
{
  "params": [
    {
      "peerId": "9444e269-007b-4f62-a87e-5d7c9a8f9cb3",
      "peerMeta": {
        "description": "An example that showcases how to use the WalletConnectSharp library",
        "url": "https://app.warriders.com/",
        "icons": [
          "https://app.warriders.com/favicon.ico"
        ],
        "name": "WalletConnectSharp"
      },
      "chainId": 1
    }
  ],
  "id": 1669119993196349,
  "jsonrpc": "2.0",
  "method": "wc_sessionRequest"
}
```

<a name="AES256CBC"/>

## AES-256-CBC PKCS7

 🟢  Generated data, using code written by me.

<b>KEY:</b> edb61c40d8bb1afbde46ea8d2aa2c86eec2ec6cf494c8e5bf9bfca9693e25d85
```json
{ 
 "iv": "ced1c84be39b2ed0086710bcf7fbaf91",
 "hmac": "958f28ad618473f02127de7acd180265291f20be7ad9335f5c692e60890fdde5", 
 "data": "bee468276df13615541444d0f19ddda01872b8eb8a6ab5385428591bad3161f88e0f41e478085a16707ea2a09f2824cee2cdf7d30314622b78b7f704b6384946d9dc492e25980c403cc40d15fb313d8abb35b5e5af0f2d5a2d7e47f23d3df1d48c64105dfbe51d07d230fb9867023f9a207441b35e55d2e6070271797c140fd431b2b879d00105639d67cfad291472007486558e70b86be8a5b0ee924b95190fd590df763cd7665ef8a942702476ca625ea9872d06d43d8078a10f590ba89e594857395a51a6337183a3fdf6803c2fce51989062c1bee0897abde2a5743e62752026d8a45c907fd2db1be0ce250da664e4607d5cb6c74938b531123dfbce71d5660294b9bddfcafa05022189b68c34d5bf931a4255282bdc326ea5e9c18098485d64cff77155e8d0038a028beb29edada8d563044fdf631820937237181b675f50f8b656a69f84725fb31d1c4167ae7896b813c2647983ec19d2d294ca0108ab8d024c3f1aa5030f8c9d58b6d0c52454"
}
```
🔵   Generated data, using code written [WalletConnectSharp](https://github.com/WalletConnect/WalletConnectSharp)

<b>KEY:</b> 8185508e7f7e4d9b2df5778a0b245a46a5f65c9b8b7184f0c7ea6de382dc0b6f
```json
{
 "iv": "fa2f3555793e088e2e7ec2adc1718cb7",
 "hmac": "0b4724d5dffd673e09b851b66d3296e1b47a591857481e5ba3f3ddf7b9fbb2ac",
 "data": "8acf555055d90f40a4197076683a0d5c547c202e21ebf39b97c623b4ca9287055f325bb1a485a619551182603b02767a5303e7c75340aa9ed20a6ff52d2a317a85f1b7cf1d8562b8b9b669496df2758ee8d3497e3ad92992470b339a5a230eea8451aecd80ef566721a427257d49f1285474324ed9eae787886f4c178a8dbacf8423ff807be529ad3695e8c55a0dcab98ee14f07ef8efdd51dd283addaaf037b4dc9d05b42da201c62eb107361699bb614898162a806b40276f85d829526db03da0878188367b469f36c2519562ef0315852c951fb4db14655d9dcdab999cfdb9d93d7d4c0e4c2dab3bd10aaa654417284c35a404e65fee924d1bac13544c62fe04798a393ca281b6444a4e4e8013479b5ab32efca9b17d94384a1a2dbffa2c264dcb9bbbd31324b5bef2660af5713d9a63e4e84ce2f3d633abf9a72570e6d050de558e4a91391d55fc2fa06d02bc6a8151c30fd46a8949e620ff43523a52436"
}
```
<a name="Publish"/>

## Publish
 🟢  Generated data, using code written by me.
```json
{
  "topic": "1fa6d81a-4888-966f-d0f9-29a3f45971ce",
  "type": "pub",
  "payload": "{\"iv\":\"ced1c84be39b2ed0086710bcf7fbaf91\",\"hmac\":\"958f28ad618473f02127de7acd180265291f20be7ad9335f5c692e60890fdde5\",\"data\":\"bee468276df13615541444d0f19ddda01872b8eb8a6ab5385428591bad3161f88e0f41e478085a16707ea2a09f2824cee2cdf7d30314622b78b7f704b6384946d9dc492e25980c403cc40d15fb313d8abb35b5e5af0f2d5a2d7e47f23d3df1d48c64105dfbe51d07d230fb9867023f9a207441b35e55d2e6070271797c140fd431b2b879d00105639d67cfad291472007486558e70b86be8a5b0ee924b95190fd590df763cd7665ef8a942702476ca625ea9872d06d43d8078a10f590ba89e594857395a51a6337183a3fdf6803c2fce51989062c1bee0897abde2a5743e62752026d8a45c907fd2db1be0ce250da664e4607d5cb6c74938b531123dfbce71d5660294b9bddfcafa05022189b68c34d5bf931a4255282bdc326ea5e9c18098485d64cff77155e8d0038a028beb29edada8d563044fdf631820937237181b675f50f8b656a69f84725fb31d1c4167ae7896b813c2647983ec19d2d294ca0108ab8d024c3f1aa5030f8c9d58b6d0c52454\"}",
  "silent": true
}
```
 🔵   Generated data, using code written [WalletConnectSharp](https://github.com/WalletConnect/WalletConnectSharp)
```json
{
  "topic": "caf40d9a-387d-4bff-8e31-0bbedefae326",
  "type": "pub",
  "payload": "{\"iv\":\"fa2f3555793e088e2e7ec2adc1718cb7\",\"hmac\":\"0b4724d5dffd673e09b851b66d3296e1b47a591857481e5ba3f3ddf7b9fbb2ac\",\"data\":\"8acf555055d90f40a4197076683a0d5c547c202e21ebf39b97c623b4ca9287055f325bb1a485a619551182603b02767a5303e7c75340aa9ed20a6ff52d2a317a85f1b7cf1d8562b8b9b669496df2758ee8d3497e3ad92992470b339a5a230eea8451aecd80ef566721a427257d49f1285474324ed9eae787886f4c178a8dbacf8423ff807be529ad3695e8c55a0dcab98ee14f07ef8efdd51dd283addaaf037b4dc9d05b42da201c62eb107361699bb614898162a806b40276f85d829526db03da0878188367b469f36c2519562ef0315852c951fb4db14655d9dcdab999cfdb9d93d7d4c0e4c2dab3bd10aaa654417284c35a404e65fee924d1bac13544c62fe04798a393ca281b6444a4e4e8013479b5ab32efca9b17d94384a1a2dbffa2c264dcb9bbbd31324b5bef2660af5713d9a63e4e84ce2f3d633abf9a72570e6d050de558e4a91391d55fc2fa06d02bc6a8151c30fd46a8949e620ff43523a52436\"}",
  "silent": true
}
```

**Comment by ecp4224** [comment, marked as answer]
Are these responses using the same secret key? When I was first writing WalletConnectSharp, I tested using the same key to ensure I was generating the same message as the typescript lib

**Reply by shkvik** [reply]
Or does Wallet Connect have an open repository with a server that accepts real requests? I could just deploy it locally and check the entire json rpc processing pipeline.

**Reply by ecp4224** [reply]
What wallet are you using to test the session connection with? Testing with WalletConnectSharp, you won't actually receive any response from the server until a wallet responses to the session connection using the URI

To test, I usually use a QR code generator site and generate a QR code for the URI. Then scan that QR code in MetaMask mobile or another mobile wallet

**Reply by ecp4224** [reply]
You can confirm this by putting a breakpoint [here](https://github.com/WalletConnect/WalletConnectSharp/blob/main/WalletConnectSharp.Desktop/Network/WebsocketTransport.cs#L73) (or adding a `Console.WriteLine`) and running the example. You'll see that you don't get any response when you initially run the example. You only get a response when you connect a wallet (or if a wallet rejects)

**Reply by shkvik** [reply]
We use Trust Wallet, it connects perfectly when the data is generated by WalletConnectSharp. We tried to send json rpc with data from WalletConnectSharp through different implementations of web sockets, all attempts were successful.

We also tried using the generated WalletConnectSharp key in generating our request. AES-256-CBC PKCS7 matches completely, hmac256 authentication is also successful.

Decryption testing took place through the method [DecryptWithKey](https://github.com/WalletConnect/WalletConnectSharp/blob/1939464b8da11a5080cb64a3435501be4f30fe06/WalletConnectSharp.Core/Network/AESCipher.cs#L62), everything was successful. But when testing with Trust Wallet, there is no response from the socket.

**Reply by shkvik** [reply]
Thank you brother!!! I did it!

**Comment by shkvik** [comment]
@ecp4224 Answer to the question: Our first message uses different keys.

Now we have tested encrypted data with the same key and iv that we took from WalletConnectSharp, all encrypted and decrypted data are completely the same.
Perhaps there are additional conditions for entry?

Are there any peculiarities or dependencies in key generation?
Perhaps you have encountered a problem that the key generation should not be random, but according to a certain encryption pattern?
